evm/src/cpu/control_flow.rs

```diff
@@ -55,10 +51,23 @@ pub fn eval_packed_generic<P: PackedField>(
     nv: &CpuColumnsView<P>,
     yield_constr: &mut ConstraintConsumer<P>,
 ) {
+    // The null special instruction must be boolean.
+    yield_constr.constraint(lv.halt_state * (lv.halt_state - P::ONES));
+    // The first row cannot be a null row. In parallel, the `bootstrap_kernel` module
+    // enforces that the first row is a kernel bootstrapping row, implicitely enforcing
+    // that the operation flags are also turned off on that row.
+    yield_constr.constraint_first_row(lv.halt_state);
```

Comment:
Reduntant constraint: Since this is already enforced in `bootstrap_kernel`, I would delete it here. There's no need for duplication.

Also, in some sense, it's okay for the first row to be `halt_state` (I think). It just means that the proof does nothing whatsoever.

Reply:
Actually you're right, I had somehow misread the way `bootstrap_kernel` was constrained on the first row..